src/zcash/IncrementalMerkleTree.hpp

```diff
@@ -0,0 +1,131 @@
+#ifndef ZCINCREMENTALMERKLETREE_H_
+#define ZCINCREMENTALMERKLETREE_H_
+
+#include <deque>
+#include <boost/optional.hpp>
+#include <boost/static_assert.hpp>
+
+#include "uint256.h"
+#include "serialize.h"
+
+static const unsigned int INCREMENTAL_MERKLE_TREE_DEPTH = 20;
+static const unsigned int INCREMENTAL_MERKLE_TREE_DEPTH_TESTING = 4;
+
+namespace libzcash {
+
+class MerklePath {
+public:
+    std::vector<std::vector<bool>> authentication_path;
+    std::vector<bool> index;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(authentication_path);
+        READWRITE(index);
+    }
+
+    MerklePath() { }
+
+    MerklePath(std::vector<std::vector<bool>> authentication_path, std::vector<bool> index)
+    : authentication_path(authentication_path), index(index) { }
+};
+
+template<size_t Depth, typename Hash>
+class IncrementalWitness;
+
+template<size_t Depth, typename Hash>
+class IncrementalMerkleTree {
+
+friend class IncrementalWitness<Depth, Hash>;
+
+public:
+    BOOST_STATIC_ASSERT(Depth >= 1);
+
+    IncrementalMerkleTree() { }
+
+    void append(Hash obj);
+    Hash root() const {
+        return root(Depth, std::deque<Hash>());
+    }
+
+    IncrementalWitness<Depth, Hash> witness() const {
+        return IncrementalWitness<Depth, Hash>(*this);
+    }
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(left);
+        READWRITE(right);
+        READWRITE(parents);
+
+        wfcheck();
+    }
+
+private:
+    boost::optional<Hash> left;
+    boost::optional<Hash> right;
+    std::vector<boost::optional<Hash>> parents;
```

Comment:
... and what it means for ``parents.size()`` to be less than ``Depth``. Or maybe it would be simpler to make it always that length (when we remove the fixpoint).